sbot/robot.py

```diff
@@ -1,189 +1,302 @@
+
+        See the metadata module for more information.
+
+        :raises MetadataNotReadyError: If the start button has not been pressed yet
+        :return: The metadata dictionary
+        """
         if self._metadata is None:
-            raise metadata.MetadataNotReadyError()
+            raise MetadataNotReadyError()
         else:
             return self._metadata
 
     @property
+    @log_to_debug
     def zone(self) -> int:
-        """The robot's starting zone in the arena (0, 1, 2 or 3)."""
-        try:
-            return cast(int, self.metadata["zone"])
-        except KeyError:
-            raise metadata.MetadataKeyError("zone") from None
+        """
+        Get the zone that the robot is in.
+
+        :return: The robot's zone number
+        :raises MetadataNotReadyError: If the start button has not been pressed yet
+        """
+        return self.metadata['zone']
 
     @property
+    @log_to_debug
     def is_competition(self) -> bool:
-        """Whether the robot is in a competition or development environment."""
-        try:
-            return cast(bool, self.metadata["is_competition"])
-        except KeyError:
-            raise metadata.MetadataKeyError("is_competition") from None
-
-    def sleep(self, secs: float) -> None:
-        """Pause the program for secs seconds."""
-        sleep(secs)
+        """
+        Find out if the robot is in competition mode.
 
-    # Custom functionality
+        :return: Whether the robot is in competition mode
+        :raises MetadataNotReadyError: If the start button has not been pressed yet
+        """
+        return self.metadata['is_competition']
 
+    @log_to_debug
     def wait_start(self) -> None:
-        """Wait for the start button to be pressed."""
-        LOGGER.info("Waiting for start button.")
-        self.power_board.piezo.buzz(timedelta(seconds=0.1), Note.A6)
-        self.power_board.wait_for_start_flash()
-        LOGGER.info("Start button pressed.")
-
-        default_metadata: Dict[str, Any] = {
-            "is_competition": False,
-            "zone": 0,
-        }
-        self._metadata = metadata.load(fallback=default_metadata)
+        """
+        Wait for the start button to be pressed.
+
+        The power board will beep once when waiting for the start button.
+        The power board's run LED will flash while waiting for the start button.
+        Once the start button is pressed, the metadata will be loaded and the timeout
+        will start if in competition mode.
+        """
+        # ignore previous button presses
+        _ = self.power_board._start_button()
+        logger.info('Waiting for start button.')
+
+        self.power_board.piezo.buzz(0.1, Note.A6)
+        self.power_board._run_led.flash()
+
+        while not self.power_board._start_button():
+            sleep(0.1)
+        logger.info("Start button pressed.")
+        self.power_board._run_led.on()
+
+        self._metadata = metadata.load()
```

Comment:
Suggestion: We've been considering `wait_start` as part of the public API. Given that, should we only load metadata if we haven't already? Save it potentially doing weird things.

Reply:
dfe0af2e39f4c9a32787f727f3da55a1285fb845